snode/src/main/java/org/apache/rocketmq/snode/SnodeStartup.java

```diff
@@ -0,0 +1,237 @@
+
+public class SnodeStartup {
+    private static InternalLogger log = InternalLoggerFactory.getLogger(LoggerName.SNODE_LOGGER_NAME);
+    public static Properties properties = null;
+    public static CommandLine commandLine = null;
+    public static String configFile = null;
+    private static final String DEFAULT_MQTT_CONFIG_FILE = "/conf/mqtt.properties";
+    private static String mqttConfigFileName = System.getProperty("rocketmq.mqtt.config", DEFAULT_MQTT_CONFIG_FILE);
+
+    public static void main(String[] args) throws IOException, JoranException {
+        SnodeConfig snodeConfig = loadConfig(args);
+        MqttConfig mqttConfig = loadMqttConfig(snodeConfig);
+        if (snodeConfig.isEmbeddedModeEnable()) {
+            BrokerController brokerController = BrokerStartup.createBrokerController(args);
+            BrokerStartup.start(brokerController);
+            snodeConfig.setSnodeName(brokerController.getBrokerConfig().getBrokerName());
+        }
+        SnodeController snodeController = createSnodeController(snodeConfig, mqttConfig);
+        startup(snodeController);
+    }
+
+    public static SnodeController startup(SnodeController controller) {
+        try {
+            controller.start();
+            String tip = "The snode[" + controller.getSnodeConfig().getSnodeName() + ", "
+                + controller.getSnodeConfig().getSnodeIP1() + "] boot success. serializeType=" + RemotingCommand.getSerializeTypeConfigInThisServer();
+
+            if (null != controller.getSnodeConfig().getNamesrvAddr()) {
+                tip += " and name server is " + controller.getSnodeConfig().getNamesrvAddr();
+            }
+            log.info(tip);
+            System.out.printf("%s%n", tip);
+            return controller;
+        } catch (Throwable e) {
+            e.printStackTrace();
+            System.exit(-1);
+        }
+        return null;
+    }
+
+    public static SnodeConfig loadConfig(String[] args) throws IOException {
+        Options options = ServerUtil.buildCommandlineOptions(new Options());
+        commandLine = ServerUtil.parseCmdLine("snode", args, buildCommandlineOptions(options),
+            new PosixParser());
+        if (null == commandLine) {
+            System.exit(-1);
+        }
+
+        SnodeConfig snodeConfig = new SnodeConfig();
+        final ServerConfig nettyServerConfig = new ServerConfig();
+        final ClientConfig nettyClientConfig = new ClientConfig();
+        nettyServerConfig.setListenPort(snodeConfig.getListenPort());
+        nettyClientConfig.setUseTLS(Boolean.parseBoolean(System.getProperty(TLS_ENABLE,
+            String.valueOf(TlsSystemConfig.tlsMode == TlsMode.ENFORCING))));
+
+        if (commandLine.hasOption('c')) {
+            String file = commandLine.getOptionValue('c');
+            if (file != null) {
+                configFile = file;
+                InputStream in = new BufferedInputStream(new FileInputStream(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *in*. The resource is referenced in statements at the following lines: 111, 113. The resource closure statement is at line: 118. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by Properties.load. Either a) close *in* in a try-finally block or b) close the resource by declaring *in* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).